clang/include/clang/Basic/Attr.td

```diff
@@ -2944,6 +2944,8 @@ def ReqdWorkGroupSize : InheritableAttr {
               ExprArgument<"YDim", /*optional*/1>,
               ExprArgument<"ZDim", /*optional*/1>];
   let Subjects = SubjectList<[Function], ErrorDiag>;
+  let Accessors = [Accessor<"usesOpenCLArgOrdering",
+                   [GNU<"reqd_work_group_size">]>];
```

Comment:
> The `[[cl::reqd_work_group_size]]` spelling uses the OpenCL arg ordering too, doesn't it? (The 1,2,1 spec defers all its attributes to the OpenCL spec, IIRC.)

That's what my understanding was- both GNU and CXX `[[cl::reqd_work_group_size]]` spelling maps to OpenCL Semantics.

According to the documentation and @mendell27, [[cl::reqd_work_group_size(Z, Y, X)]] attribute should be Z,Y,X for SYCL (same as [[intell::reqd_work_group_size(Z, Y, X)]] attribute) . 
https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide/top/optimize-your-design/resource-use/specify-a-work-group-size.html

Reply:
@gmlueck -- do you have an opinion?

Reply:
tagging @MrSidims and @GarveyJoe

Reply:
Since `[[cl::reqd_work_group_size()]]` is deprecated and not well specified, we should do whatever will break the least amount of code.  Does DPPC++ already implement this deprecated attribute?  If so, how are the arguments currently interpreted?

That said, after chatting with @mkinsner and @Pennycook, we think the intent of 1.2.1 was to interpret the parameters as (z, y, x), where x varies fastest.  Hopefully, this is how DPC++ currently interprets them (if it's currently implemented).

Reply:
Thanks @gmlueck for comments.

>>Does DPPC++ already implement this deprecated attribute?

yes, ``[[cl::reqd_work_group_size()]] has been implemented. 

https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/AttrDocs.td#L2575

Reply:
@gmlueck -- how did you determine that intent from the spec? I can't see any way to read the 1.2.1 spec that suggests anything *but* following the same semantics as OpenCL.

> The attribute syntax defined in the OpenCL C specification is supported by the SYCL device compiler. The
C++11 attribute specifier can be used in which case these attributes are available in the cl namespace. For instance
__attribute__((vec_type_hint(int2))) and [[cl::vec_type_hint(int2)]] are equivalent.

If `__attribute__((vec_type_hint(int2)))` and `[[cl::vec_type_hint(int2)]]` are equivalent, why would it be different for `reqd_work_group_size`?

Reply:
I should be clear about my concern -- I'm worried that "do whatever breaks less code" will lead to more bug reports and more difficulty upstreaming because anyone reading the standard will come to the wrong conclusion about what the code does. We already have a hard enough time with `reqd_work_group_size` behaving mysteriously, so I worry deviating from the standard will cause additional user confusion.

Reply:
Let me start by saying that we all feel the SYCL 1.2.1 spec is very unclear about the meaning of the attribute parameters, and we think new code should use the new attributes in the `sycl::` namespace, which are much better defined.  It probably would have been better to either clarify the parameters of the deprecated attributes in SYCL 2020 or just remove them entirely from the spec.  However, that is not what happened.  They are just as unclear in SYCL 2020 as they were in 1.2.1.

I agree that the section of the 1.2.1 spec describing attributes leads the reader to believe that they should probably follow the OpenCL ordering.  However, section 4.8.10 "OpenCL Kernel Conventions and SYCL" of the 1.2.1 spec clearly says that the dimension order in SYCL is opposite of that in OpenCL.  That section even includes the following general statement about dimension ordering:

> The mapping of highest dimension in SYCL to lowest dimension in OpenCL applies to all operations where
a multi-dimensional construct must be mapped ...

I believe this is why @mkinsner and @Pennycook believe the intent was (z, y, x).  (I was not involved with SYCL 1.2.1, so I have no first hand knowledge.)

I think everyone in the language evolution team wants to avoid breaking existing code, though.  If the DPC++ compiler currently implements a particular semantic, it seems better to continue using that semantic, especially since the 1.2.1 spec is so unclear.  If customers report bugs, our answer should be to use the non-deprecated attribute instead.

Reply:
Thanks for clarifying, @gmlueck. I think I may have different expectations about what "doesn't want to break existing code" means. When there's a standard and an implementation of the standard, I expect any disagreement between the standard and the implementation to be resolved in favor of the standard unless what the standard says is obviously a bug (e.g., it would be impossible for the implementation to do what the standard says). Otherwise, the standard isn't really standardizing anything because code isn't portable between implementations, which is a form of code breakage (conforming code can't be compiled properly by our implementation). I think you're viewing code breakage as purely about how the compiler currently behaves, which is a different approach. We definitely disagree on how to read the SYCL 1.2.1 standard in this case, so I think there's the potential for code breakage regardless of what way we resolve this.

That said, I think the best answer here is to treat `[[cl::reqd_work_group_size]]` as a synonym for `[[intel::reqd_work_group_size]]` rather than `__attribute__((reqd_work_group_size))` in terms of its semantics, but deprecate any use of that spelling in SYCL 2020 mode. That deprecation work will happen as a part of https://github.com/intel/llvm/pull/3689/.

We'll have to treat the "optional-but-not-actually" arguments the same for `[[cl::reqd_work_group_size]` and `[[intel::reqd_work_group_size]]` as well, but can address that mess in a separate follow-up.

Reply:
> I agree that the section of the 1.2.1 spec describing attributes leads the reader to believe that they should probably follow the OpenCL ordering. However, section 4.8.10 "OpenCL Kernel Conventions and SYCL" of the 1.2.1 spec clearly says that the dimension order in SYCL is opposite of that in OpenCL. That section even includes the following general statement about dimension ordering:

Wait a second, now I guess I need to verify something else... are you suggesting that `__attribute__((reqd_work_group_size))` is *also* supposed to be following the SYCL 2020 ordering, not just [[cl::reqd_work_group_size]]`?

Reply:
We've been thinking about this more, and we think it might be easier for everyone if we agree to remove the old spelling of these attributes from the SYCL 2020 spec, rather than just deprecating them.  As we've been discussing, they are very poorly defined at present, and we really just want people to stop using them and shift to the new ones in the `sycl::` namespace.  I'll raise this with the Khronos group next week and see if there is an appetite for removing them as a DR fix.  If we do not agree to remove them, I'll try to push for defining them more clearly in the spec.

Answering your specific question, though ... I would not call it the "SYCL 2020 ordering" because section 4.8.10 of the SYCL 1.2.1 spec also says that the SYCL ordering is reverse from the OpenCL ordering.  Therefore, it's really the "SYCL ordering", not the "SYCL 2020 ordering".  However, yes, the order of the parameters to both `__attribute__((reqd_work_group_size))` and `[[cl::reqd_work_group_size]]` are the same, both reversed from OpenCL.  For that matter, the order of the parameters to `work_group_size_hint` should also match the SYCL way.

We realize that the GNU spelling (`__attribute__((reqd_work_group_size))`) is exactly the same as the OpenCL attribute, which interprets the parameters in the opposite order.  Therefore, the compiler will need to interpret the arguments differently depending on whether it is consuming SYCL code vs. OpenCL code.

Reply:
>> However, yes, the order of the parameters to both __attribute__((reqd_work_group_size)) and [[cl::reqd_work_group_size]] are the same, both reversed from OpenCL. 

Thanks @gmlueck. You are suggesting that all three current spellings of ReqdWorkGroupSizeAttr (```__attribute__((reqd_work_group_size)), [[cl::reqd_work_group_size]], and [[intel::reqd_work_group_size]]``) is supposed to follow SYCL argument ordering (Z, Y, X) even though GNU spelling (__attribute__((reqd_work_group_size))) is exactly the same as the OpenCL attribute.

Before i update my current PR, i want to make sure that i understand correctly.

Reply:
@gmlueck -- Thanks for the further information. I'm glad I asked, because we were about to do the wrong thing again. :-) FWIW, I'd be in favor of removing `reqd_work_group_size` spellings from SYCL 2020 mode, but I want to make sure I'm on the same page. I read that suggestion as meaning: `__attribute__((reqd_work_group_size))` is only available in OpenCL mode and follows the OpenCL rules. `[[sycl::reqd_work_group_size]]`, `[[cl::reqd_work_group_size]]`, and `[[intel::reqd_work_group_size]]` are only available in SYCL modes and follow the SYCL rules. The `__attribute__((reqd_work_group_size))` and `[[cl::reqd_work_group_size]]` spellings get an unknown attribute warning if used in SYCL 2020 mode and the `[[sycl::reqd_work_group_size]]` spelling gets an extension warning if used in SYCL 2017 mode.

@smanna12 -- I read the suggestion (assuming we *do not* remove the attribute support in SYCL 2020 mode) as all spellings of `reqd_work_group_size` (regardless of syntax used) follow the SYCL rules when in SYCL mode.

Reply:
>>@smanna12 -- I read the suggestion (assuming we do not remove the attribute support in SYCL 2020 mode) as all spellings of reqd_work_group_size (regardless of syntax used) follow the SYCL rules when in SYCL mode.

Thanks @AaronBallman, Yes, All spellings are following SYCL rules in SYCL mode. I have updated current PR.

>>@gmlueck -- Thanks for the further information. I'm glad I asked, because we were about to do the wrong thing again. :-) 

 yes, i agree.
My understanding was that GNU spelling will always follow OpenCL argument orders in both OpenCL and SYCL mode. 
Thank you @AaronBallman and @gmlueck for the clarification.

Reply:
> I read that suggestion as meaning: `__attribute__((reqd_work_group_size))` is only available in OpenCL mode and follows the OpenCL rules. `[[sycl::reqd_work_group_size]]`, `[[cl::reqd_work_group_size]]`, and `[[intel::reqd_work_group_size]]` are only available in SYCL modes and follow the SYCL rules. The `__attribute__((reqd_work_group_size))` and `[[cl::reqd_work_group_size]]` spellings get an unknown attribute warning if used in SYCL 2020 mode and the `[[sycl::reqd_work_group_size]]` spelling gets an extension warning if used in SYCL 2017 mode.

This is my suggestion for SYCL 2020.  However, for SYCL 1.2.1, I think we still need to support `__attribute__((reqd_work_group_size))`, using the SYCL rules for argument order.

Note also that this suggestion is dependent on the Khronos group agreeing with me, which I don't know yet.